### PR TITLE
Source File: Add option to Generic Zipped Files

### PR DESCRIPTION
**Pull request by madhup**
## What
Extends the file source to support zipped/archive files containing multiple files
https://github.com/airbytehq/hacktoberfest-2021/issues/63

## How
Creates a wrapper around file-source. If the provided file is an archive file, all the individual files are extracted locally and treated as individual stream. Otherwise, the original files workflow is followed.
   
#### Community member or Airbyter
   
- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-file/setup.py [hunk 1]
```diff
@@ -28,7 +28,7 @@
 setup(
     name="source_file",
     description="Source implementation for File",
-    author="Airbyte",
+    author="Madhup Sukoon",
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-file/setup.py:31]: Revert this.

Comment (vagrantism) [reply airbyte-integrations/connectors/source-file/setup.py]: Done!

**Review by alafanechere — CHANGES_REQUESTED** [review]
Hi @vagrantism thanks for your contribution.
I still have some requirements before my first full review:
* Could you please share evidence that the integration tests are passing?
* Could you go through the premerge checklist in description? 
* It looks like you added a dependency (`numpy`) that is not declared in the `setup.py`
* I see you're removing the usage of the original `Client` class and replacing it by `ArchiveClient` but keeping both classes. I'm not sure the original use case with non archived files is still working.
* You're tweaking the `spec.json` a lot, and again I'm not sure that the original use case for this connector is still covered, have you thought of creating a specific connector for the archives?

**Comment by vagrantism** [comment]
Hey @alafanechere , hope these help:

> Could you please share evidence that the integration tests are passing?
I could not run the original integration tests as the `source-files` connector [requires access to some cloud providers](https://github.com/airbytehq/airbyte/tree/master/airbyte-integrations/connectors/source-file#necessary-credentials-for-tests) and I do hot have these credentials. However, I have tested the connector with it's actual use-case (directly importing files) as well as the extended use case (importing zipped/archive files)

> Could you go through the premerge checklist in description?
Done, but I'll need help updating the documentation.

> It looks like you added a dependency (numpy) that is not declared in the setup.py
Numpy is a dependency for Pandas, which was already being used. If you need me to explicitly include Numpy in setup.py, let me know and I'll do it.

> I see you're removing the usage of the original Client class and replacing it by ArchiveClient but keeping both classes. I'm not sure the original use case with non archived files is still working. You're tweaking the spec.json a lot, and again I'm not sure that the original use case for this connector is still covered, have you thought of creating a specific connector for the archives? 

The original use case is being covered, and that's the reason for the changes in spec.json - I modified it so that the user can specify whether the remote file should be considered as a zipped/archive fire, or a regular, direct file - allowing for both workflows. I initially developed this as a separate connector, but refactored it to extend `file-source` instead  as asked [here](https://github.com/airbytehq/hacktoberfest-2021/issues/63#issuecomment-950586817)

**Comment by vagrantism** [comment]
Hey @alafanechere, any updates?

**Comment by marcosmarxm** [comment]
Hello @vagrantism ! Sorry to not finished your contribution before the date stipulated in the contest. All contributions made before 15- November are eligible to receive the award. We're trying to review your contribution as soon as possible.

File: airbyte-integrations/connectors/source-file/source_file/spec.json [hunk 2]
```diff
@@ -6,22 +6,31 @@
...
+        "default": "default_tablename",
+        "examples": ["single_file_table_name", "{\"file1.xyz\": \"tablename1\", \"file2.xyz\": \"tablename2\"}"]
       },
       "reader_options": {
         "type": "string",
+        "default": "{}",
         "description": "This should be a valid JSON string used by each reader/parser to provide additional options and tune its behavior",
-        "examples": ["{}", "{'sep': ' '}"]
+        "examples": ["{\"all\": {\"sep\": \" \"}}", "{\"file1.xyz\": {\"sep\": \" \"}, \"file2.xyz\": {\"sep\": \"|\"}}"]
       },
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-file/source_file/spec.json:34]: I think that tweaking the specs as you did leads to a less strict validation of parameters that can lead to more error for users.
What do you think of changing the spec structure to have a more user friendly interface such as:
```json
{
    "documentationUrl": "https://docs.airbyte.io/integrations/sources/file",
    "connectionSpecification": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "title": "File Source Spec",
        "type": "object",
        "additionalProperties": false,
        "required": ["url", "provider", "is_archive", "files"],
        "properties": {
            "url": {
                "type": "string",
                "description": "URL path to access the file to be replicated"
            },
            "is_archive": {
              "type": "boolean",
              "description": "Wether the source file is a tar archive"
            },
            "provider": {},
            "files": {
                "type": "array",
                "description": "List of files to download, set multiple files if the source is an archive",
                "items": {
                    "type": "object",
                    "additionalProperties": false,
                    "required": [
                        "format",
                        "dataset_name"
                    ],
                    "properties": {
                        "format": {
                            "type": "string",
                            "enum": ["csv", "json", "jsonl", "excel", "feather", "parquet"
                            ],
                            "default": "csv",
                            "description": "File Format of the file to be replicated (Warning: some format may be experimental, please refer to docs)."
                        },
                        "dataset_name": {
                            "type": "string",
                            "description": "Name of the final table where to replicate this file (should include only letters, numbers dash and underscores)"
                        },
                        "reader_options": {
                            "type": "string",
                            "description": "This should be a valid JSON string used by each reader/parser to provide additional options and tune its behavior",
                            "examples": ["{}", "{'sep': ' '}"]
                        }
                    }
                }
            }
        }
    }
}
```

File: airbyte-integrations/connectors/source-file/source_file/client.py [hunk 3]
```diff
@@ -363,3 +367,126 @@ def streams(self) -> Iterable:
...
+                    }
+            else:
+                raise NotImplementedError
+
+
+    @property
+    def streams(self) -> Iterable:
+        """Discovers available streams"""
+
+        dtype_map = {
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-file/source_file/client.py:461]: You already have this mapping in `Client.dtype_to_json_type`

File: airbyte-integrations/connectors/source-file/source_file/client.py [hunk 4]
```diff
@@ -363,3 +367,126 @@ def streams(self) -> Iterable:
...
+    def reader(self) -> reader_class:
+        return self.reader_class(url=self._url, provider=self._provider)
+
+    @property
+    def binary_source(self):
+        binary_formats = {"excel", "feather", "parquet", "orc", "pickle"}
+        return (
+            (self._archive_format != 'Not Archived') or 
+            (self._file_format['all'] in binary_formats)
+        )
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-file/source_file/client.py:424]: These properties are already declared in Client class. You could leverage inheritance.

File: airbyte-integrations/connectors/source-file/source_file/client.py [hunk 5]
```diff
@@ -363,3 +367,126 @@ def streams(self) -> Iterable:
...
+
+    def extract(self) -> Mapping[str, Client]:
+        if self._archive_format == 'Not Archived':
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-file/source_file/client.py:427]: I think extract should not be called if the source file is not an archive

File: airbyte-integrations/connectors/source-file/source_file/source.py [hunk 6]
```diff
@@ -67,7 +67,7 @@ class SourceFile(Source):
     content of directories, etc in a latter iteration.
     """
 
-    client_class = Client
+    client_class = ArchiveClient
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-file/source_file/source.py:70]: This is a bit misleading, at first read it suggests that all the file are considered to be archive. I'd prefer to have a kind of factory method (`_get_client`?)  which creates archive client or normal file client.

**Review by alafanechere — CHANGES_REQUESTED** [review]
Hi @vagrantism I made my first high level review.
My main concerns are the following: 
* I find that your spec file modification leads to less strict validation. Our file connector is a popular one and we'd love to minimize error related to user inputs. I made a spec structure suggestion, I'd love feedback on this.
* There's a lot of duplicate code between `Client` and `ArchiveClient`. I'd love to see more inheritance and different explicit processing path for archive / normal file. I find the client module a bit hard to read after your modifications.

About acceptance tests: It might be hard indeed to run all the tests. But in the current branch state, the checks made before the integration test are failing, you need to be sure that the `flakeCheck` is passing correctly and it does not require you to have access to cloud providers. You could also locally tweak the integration tests to discard provider and only check scp/ssh/sftp. I'll take care of running the integration test for the cloud providers you do not have access to.

**Comment by marcosmarxm** [comment]
I'll add the label `blocked` until the user reply the review.

**Comment by marcosmarxm** [comment]
@vagrantism  I'll close this due to inactivity. If you want to return this work you can reopen the PR.

**Pull request closed without merging**